source/extensions/filters/common/expr/custom_cel/custom_cel_vocabulary.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+
+#include "source/common/protobuf/protobuf.h"
+#include "source/extensions/filters/common/expr/context.h"
+
+#include "absl/strings/string_view.h"
+#include "eval/public/activation.h"
+#include "eval/public/cel_function_registry.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Expr {
+namespace Custom_CEL {
+
+using google::api::expr::runtime::Activation;
+using google::api::expr::runtime::CelFunctionRegistry;
+
+// CustomCELVocabulary
+//
+// Background Information:
+//
+// CelExpression: CelExpressions can contain values, variables and functions.
+// The functions can be (1) lazy functions or (2) static (eagerly evaluated) functions.
+// Lazy functions are evaluated during the request.
+// Static functions are evaluated once during the duration of the application.
+//
+// CelExpressions are evaluated with the aid of certain data structures
+// which contain references to variables and functions implementations.
+// The data structures are (1) the CelFunctionRegistry and (2) the Activation.
+//
+// CelFunctionRegistry: A registry of all CEL functions which may be used in
+// CelExpressions, and which lasts for the life of the application.
+// The registry contains registrations for both lazy functions and static functions.
+//
+// Activation: contains a mapping of function names and variable names to their
+// reference implementations.
+// A new activation is created for every request and lasts for the duration of
+// a single request.
+// An activation contains lazy functions but not static functions.
+//
+// Vocabulary: The variables and functions constitute the "vocabulary" used to
+// evaluate expressions.
+//
+// There are multiple vocabularies in use:
+//
+// (1) the Built-in CEL Vocabulary
+// https://github.com/google/cel-spec/blob/master/doc/langdef.md#list-of-standard-definitions
+// Built-in CEL Functions include "+", "-", "*".
+// These are static and registered with the CEL function registry.
```

Comment:
If these are static, how are they only evaluated on ceduring the application lifetime? Surely we evaluate `+` for every new set of args in an expression?

Reply:
Yes.  This was a serious mistake.  I will correct the comments.

Reply:
It should have been: 

There is a CelExpressionBuilder which can produce CelExpressions.  
The CelExpressionBuilder has a CEL function registry of all allowable CEL functions.  
An activation is created for every CelExpression.  There is one registry for every set of CelExpressions (produced by a single CelExpressionBuilder). 

The distinction is between stateful and stateless functions.  CEL calls them lazy (stateful) and static (stateless).